File: DMC/updates/changeInternalPhononMomentumMagnitude.py
import sys
import numpy as np
from scipy.special import erfinv, erf

sys.path.append('/dependencies')
from .dependencies.calculateQ import calculateQ
from .dependencies.calculateP0 import calculateP0

def changeInternalPhononMomentumMagnitude(self):
  # choose phonon propagator on random
  d = np.random.choice(self.FD.Ds)

  P0 = calculateP0(d)
  p0 = np.linalg.norm(P0)

  sqrta = (0.5*(d.end.position - d.start.position))**0.5
  b = p0*np.cos(d.theta)
  r = np.random.rand()
  q = b + erfinv(r + (r - 1)*erf(sqrta*b))/sqrta

  Q = calculateQ(P0, q, d.theta, d.phi)

  if self.debug:
    qOld = np.linalg.norm(d.momentum)
    QOld = calculateQ(P0, qOld, d.theta, d.phi)
    self.FD.setInternalPhononMomentum(d, QOld)
    diagOld = self.FD()

  # update diagram
  self.FD.setInternalPhononMomentum(d, Q)

  if self.debug:
    diag = self.FD()

    R = diag/diagOld
    R *= np.exp(-sqrta**2 * ((qOld - b)**2 - (q - b)**2))

    print('Momentum Magnitude', R)

  return 1

File: c++/DMC/updates/changeInternalPhononMomentumMagnitude.cpp
#include "../DiagrammaticMonteCarlo.h"

double DiagrammaticMonteCarlo::changeInternalPhononMomentumMagnitude () {
  // select internal phonon on random
  shared_ptr<Phonon> d = this->FD.Ds[this->Uint(0, this->FD.Ds.size() - 1)];

  Vector3d P0 = this->calculateP0(d);
  double p0 = P0.norm();

  double
    param1 = sqrt(0.5*(d->end->position - d->start->position)),
    param2 = p0*cos(d->theta);

  double
    r = this->Udouble(0, 1),
    q = param2 + boost::math::erf_inv(r + (r - 1)*erf(param1*param2))/param2;

  Vector3d Q = this->calculateQ(P0, q, d->theta, d->phi);

  double oldq, oldVal;
  Vector3d oldQ;
  if (this->debug) {
    oldq = d->momentum.norm();
    oldQ = this->calculateQ(P0, oldq, d->theta, d->phi);
    this->FD.setInternalPhononMomentum(d, oldQ);
    oldVal = this->FD();
  }

  // update diagram
  this->FD.setInternalPhononMomentum(d, Q);

  if (this->debug) {
    double a, val = this->FD();

    a = val/oldVal;
    a *= exp(-pow(param1, 2) * (pow(oldq - param2, 2) - pow(q - param2, 2)));

    cout << "changeInternalPhononMomentumMagnitude: " << a << endl;
  }
  
  return 1;
}